src/tests/profiler/native/releaseondetach/releaseondetach.cpp

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "releaseondetach.h"
+
+#ifdef WIN32
+#include <Windows.h>
+#else // WIN32
+#include <dlfcn.h>
+#include <iostream>
+#include <fstream>
+#include <string>
+using std::string;
+using std::ifstream;
+using std::getline;
+#ifdef __APPLE__
+#include <mach-o/dyld.h>
+#endif // __APPLE__
+#endif // WIN32
+
+ReleaseOnDetach *ReleaseOnDetach::Instance;
+
+ReleaseOnDetach::ReleaseOnDetach() :
+    _dispenser(NULL),
+    _failures(0),
+    _detachSucceeded(false),
+    _doneFlag(NULL)
+{
+    ReleaseOnDetach::Instance = this;
+}
+
+ReleaseOnDetach::~ReleaseOnDetach()
+{
+    if (_doneFlag != NULL)
+    {
+        *_doneFlag = true;
```

Comment:
It would be better if:
+ this was set using an atomic operation to guarantee memory ordering on arm/arm64.
+ set after the `printf`s  completed.  maybe add a `fflush(stdout)` too

Reply:
What is the concern for the memory ordering on arm? I'm happy to change it if there's an issue, I just don't understand what the potential issue is.

Since the code that reads it is sitting in a spin loop, wouldn't the exact order of operations not matter? I.e. if it races with another thread it will just read it again on the next iteration?

Reply:
It is just that done is set before the `printf` (so we are not quite done).  The spin loop could possibly terminate the test before the printing of the result completed.  I am not sure how quickly it could/would force the shutdown.  It is simply my defensive programming instinct.

Reply:
Oh, ok. I see, you're concerned about whether the success message gets printed to the console. I wasn't thinking about it from that angle.

As far as I know there isn't a cross platform InterlockedExchange equivalent. I marked it as volatile, added the fflush and moved it to the end of the method. I think that's all I can do without a lot of effort.

Reply:
> As far as I know there isn't a cross platform InterlockedExchange equivalent

If our compilers are supporting at least C++ 11 (I think they do?) then we should have access to std::atomic
https://docs.microsoft.com/en-us/cpp/standard-library/atomic-functions?view=vs-2019#atomic_compare_exchange_strong

Reply:
@noahfalk atomic works if you want an int/whatever built in type with atomic semantics but it doesn’t have a way to pass it an arbitrary pointer, does it? The bool * is coming from managed code. I could refactor and have an atomic bool in the profiler that gets passed to managed code, but I just don’t think it’s worth refactoring the test again.